rl/memory.py

```diff
@@ -45,52 +67,83 @@ def __init__(self, limit):
         self.terminals = RingBuffer(limit)
         self.observations = RingBuffer(limit)
 
-    def sample(self, batch_size, window_length):
-        # Draw random indexes such that we have at least `window_length` entries before each index.
-        batch_idxs = np.random.random_integers(window_length, self.nb_entries - 1, size=batch_size)
+    def sample(self, batch_size, window_length, batch_idxs=None):
+        if batch_idxs is None:
+            # Draw random indexes such that we have at least a single entry before each
+            # index.
+            batch_idxs = sample_batch_indexes(0, self.nb_entries - 1, size=batch_size)
+        batch_idxs = np.array(batch_idxs) + 1
+        assert np.min(batch_idxs) >= 1
+        assert np.max(batch_idxs) < self.nb_entries
         assert len(batch_idxs) == batch_size
 
         # Create experiences
         experiences = []
         for idx in batch_idxs:
+            terminal0 = self.terminals[idx - 2] if idx >= 2 else False
+            while terminal0:
+                # Skip this transition because the environment was reset here. Select a new, random
+                # transition and use this instead. This may cause the batch to contain the same
+                # transition twice.
+                idx = sample_batch_indexes(1, self.nb_entries, size=1)[0]
+                terminal0 = self.terminals[idx - 2] if idx >= 2 else False
+            assert 1 <= idx < self.nb_entries
+
             # This code is slightly complicated by the fact that subsequent observations might be
             # from different episodes. We ensure that an experience never spans multiple episodes.
             # This is probably not that important in practice but it seems cleaner.
             state0 = [self.observations[idx - 1]]
-            for current_idx in range(idx - 2, idx - window_length - 1, -1):
-                if self.terminals[current_idx]:
+            for offset in range(0, window_length - 1):
```

Comment:
This seems to be the same logic as in get get_recent_state(). Combine into a common function get_state() ?